core: add a procedure to recover mon quorum from OSDs (#7154)
rook/rook: Documentation/ceph-disaster-recovery.md

> The operator will automatically add more mons to increase the quorum size again, depending on the `mon.count`.

Reviewer: There are enough recovery sections in this document, could you add links at the top of the document to the sections, similar to the links in the [ceph-common-issues.md](https://rook.github.io/docs/rook/v1.5/ceph-common-issues.html#table-of-contents-)?

> ## Recovering Mon Quorum using OSDs

Reviewer: This is a super complex scenario. I wonder how we could maintain these instructions in the long term so they will continue to work. With my work in #3895 hopefully we can set a pattern for automating more of these actions, and add tests for them. This PR isn't blocked on that, but just thinking about the long term...

Author: > hopefully we can set a pattern for automating more of these actions, and add tests for them.

I'll do after merging this PR. Is it OK to be a part of the integration test? For example, creating one OSD cluster, corrupt mon, and restore the cluster with the steps describe in this document.

> With my work in #3895

I can't interplet what doesnyour work in #3895 mean. It looks a PR for resetting a release note.

Reviewer: Typo! The issue is #3985.

If all MONs fail, we can recover monitor store from OSDs. The steps in this section are based on the following document.

https://docs.ceph.com/en/latest/rados/troubleshooting/troubleshooting-mon/#recovery-using-osds

Known limitations are written in the following document.

https://docs.ceph.com/en/latest/rados/troubleshooting/troubleshooting-mon/#known-limitations

### Prerequisites

1. OSDs are not corrupted.
2. The rook [toolbox](./ceph-toolbox.md) is deployed in the target cluster's namespace.

### Set variables related to the environment

```console

> FSID="<The value after --fsid is written in the mon deployment manifest.>"

Reviewer: This is a lot of scripting in documentation. What if we provide this as one or more bash scripts in the `test/scripts` folder? The documentation then can focus on describing the steps, and the details can be left in the script.

Reviewer 2: +1

The operator will automatically add more mons to increase the quorum size again, depending on the `mon.count`.

## Restoring Mon Quorum using healthy OSDs

If all mons fail, it might be possible to recover the monstore from the OSDs. [Ceph's documentation](https://docs.ceph.com/en/latest/rados/troubleshooting/troubleshooting-mon/#recovery-using-osds) describes how to do it.

> **WARNING**: Please keep in mind that recovery is not guaranteed for the cluster or all data, even if these steps succeed. We recommend restoring your data from backup if available.

### Prerequisites

[toolbox](./ceph-toolbox.md) is deployed.

### Preparation

* Take backup of deployments of mons, osds, and toolbox.
* Stop operator

```console

> NS=<cluster's namespace such as "rook-ceph">

Reviewer: ```suggestion
# Change to your rook namespace if different
NS=rook-ceph
```

> kubectl -n ${NS} scale deployment rook-ceph-operator --replicas=0`

Reviewer: ```suggestion
kubectl -n ${NS} scale deployment rook-ceph-operator --replicas=0
```

```

* Suspend all mon pods and osd pods as follows.

```console
MONS=<mon's deployments>
OSDS=<osd's deployments>

for mon in ${MONS}; do

>     kubectl -n ${NS} patch deployment ${mon} -p '{"spec": {"template": {"spec": {"containers": [{"name": "mon", "command": ["sleep", "infinity"], "args": []}]}}}}'

Reviewer: This line gives an error:
```
Error from server (BadRequest): invalid character ']' after object key
```

Author: I guess that I made a mistake when copy-and paste the these steps. I'll check it again later.

>     kubectl -n ${NS} patch deployment "{mon} --type=json --patch='[{"op":"remove", "path":"/spec/template/spec/containers/0/livenessProbe"}]'

Reviewer: ```suggestion
    kubectl -n ${NS} patch deployment ${mon} --type=json --patch='[{"op":"remove", "path":"/spec/template/spec/containers/0/livenessProbe"}]'
```

The operator will automatically add more mons to increase the quorum size again, depending on the `mon.count`.

## Restoring Mon Quorum using healthy OSDs

> If all mons fail, it might be able to recover monstore from OSDs. [Ceph's document](https://docs.ceph.com/en/latest/rados/troubleshooting/troubleshooting-mon/#recovery-using-osds) describes how to do it.

Reviewer: ```suggestion
If all mons fail, it might be possible to recover the monstore from the OSDs. [Ceph's documentation](https://docs.ceph.com/en/latest/rados/troubleshooting/troubleshooting-mon/#recovery-using-osds) describes how to do it.
```

    kubectl -n rook-ceph patch deployment "${mon}" --type=json --patch='[{"op":"remove", "path":"/spec/template/spec/containers/0/livenessProbe"}]'
done

for osd in ${OSDS}; do
    kubectl -n rook-ceph patch deployment "${osd}" -p '{"spec": {"template": {"spec": {"containers": [{"name": "osd", "command": ["sleep", "infinity"], "args": []}]}}}}'
    kubectl -n rook-ceph patch deployment "${osd}" --type=json --patch='[{"op":"remove", "path":"/spec/template/spec/containers/0/livenessProbe"}]'
done
```

* Inject mon-keyring to toolbox for the subsequent steps

```console
kubectl -n rook-ceph patch deployment rook-ceph-tools --type=json \
 --patch='[
     {"op":"add", "path":"/spec/template/spec/volumes/-", "value": {"name": "mons-keyring", "secret": {"secretName": "rook-ceph-mons-keyring"}}},
     {"op":"add", "path":"/spec/template/spec/containers/0/volumeMounts/-", "value": {"name": "mons-keyring", "mountPath": "/keyrings"}}
 ]'

kubectl -n rook-ceph-tools delete pod -l app=rook-ceph-tools
```

### Collect monmap from healthy osds

```console

> OSD_IDS=<healthy osd's ids like "0 1 2">

Reviewer: ```suggestion
OSD_IDS=<healthy osd's ids separated by spaces such as "0 1 2">
```

> TMP_MON_STORE=<temporal monstore's file>

Reviewer: What is an example of this monstore file?

    kubectl exec -i -n rook-ceph ${name} -- ceph-objectstore-tool --data-path /var/lib/ceph/osd/ceph-"${i}" --no-mon-config --op update-mon-db --mon-store-path "${TMP_MON_STORE}"
    rm -rf "${TMP_MON_STORE}"/*
    kubectl exec -i -n rook-ceph "${name}" -- tar cf - -C "${TMP_MON_STORE}" . | tar xf - -C "${TMP_MON_STORE}"
done
```

### Rebuild monstore

```console
FSID=$(kubectl -n rook-ceph get secret rook-ceph-mon -o jsonpath='{.data.fsid}' | base64 -d)
kubectl exec -i -n rook-ceph deployment/rook-ceph-tools -- rm -rf "${TMP_MON_STORE}"
kubectl exec -i -n rook-ceph deployment/rook-ceph-tools -- mkdir -p "${TMP_MON_STORE}"
tar cf - -C "${TMP_MON_STORE}" . | kubectl exec -i -n rook-ceph deployment/rook-ceph-tools -- tar xf - -C "${TMP_MON_STORE}"
kubectl exec -i -n rook-ceph deployment/rook-ceph-tools -- ceph-monstore-tool --fsid="${FSID}" "${TMP_MON_STORE}" rebuild -- --keyring /keyrings/keyring --mon-ids ${MON_IDS}
rm -rf "${TMP_MON_STORE}"
mkdir -p "${TMP_MON_STORE}"
kubectl exec -i -n rook-ceph deployment/rook-ceph-tools -- tar cf - -C "${TMP_MON_STORE}" . | tar xf - -C "${TMP_MON_STORE}"
```

> ### Replace the corrupted monstore with the recovered copy.

Reviewer: suggestion: keep the title short, and then give a sentence for more details
```suggestion
### Replace the corrupted monstore

Replace the corrupted monstore with the recovered copy.
```

```console
MON_IDS=<mon's ids, like "a b c">

for i in ${MON_IDS}; do
    name=$(kubectl get -n rook-ceph pod -l app=rook-ceph-mon -l mon="${i}" -o json | jq -r '.items[].metadata.name')
    STORE_DB=/var/lib/ceph/mon/ceph-"${i}"/store.db
    kubectl exec -i -n rook-ceph "${name}" -- rm -rf "${STORE_DB}".corrupted
    kubectl exec -i -n rook-ceph "${name}" -- mkdir -p "${STORE_DB}"
    kubectl exec -i -n rook-ceph "${name}" -- mv "${STORE_DB}" "${STORE_DB}".corrupted
    kubectl exec -i -n rook-ceph "${name}" -- mkdir -p "${STORE_DB}"
    tar cf - -C "${TMP_MON_STORE}"/store.db . | kubectl exec -i -n rook-ceph "${name}" -- tar xf - -C "${STORE_DB}"
    kubectl exec -i -n rook-ceph "${name}" -- chown -R ceph:ceph "${STORE_DB}"
done
```

> ### Aftertreatment

Reviewer: ```suggestion
### Restore the daemons
```

> * Restore deployments of mons, osds, and toolbox.

Reviewer: ```suggestion
* Restore the deployments for the mons, osds, and toolbox that were removed or modified in the earlier steps.
```

> * Restart the cluster

Reviewer: ```suggestion
* Restart the operator
```

```console
kubectl exec -i -n rook-ceph deployment/rook-ceph-tools -- ceph mon --connect-timeout 60 enable-msgr2
kubectl -n rook-ceph scale deployment rook-ceph-operator --replicas=1
```

> * Restart operator by `kubectl -n rook-ceph delete pod -l app=rook-ceph-operator`

Reviewer: Why restart the operator right after it was just scaled up? Maybe we don't need this step?

> * Restart mgr by `kubectl -n rook-ceph delete pod -l app=rook-ceph-mgr`

Reviewer: Why do we need to restart the mgr?

Author: Mgr doesn't work correctly and becomes CLBO after restoring the cluster probably due to missing some data. Restarting mgr resolves this problem.

The operator will automatically add more mons to increase the quorum size again, depending on the `mon.count`.

## Restoring Mon Quorum using OSDs

If all MONs fail, we can recover monitor store from OSDs. The steps in this section are based on [this document](https://docs.ceph.com/en/latest/rados/troubleshooting/troubleshooting-mon/#recovery-using-osds).

Known limitations are written in [this document](https://docs.ceph.com/en/latest/rados/troubleshooting/troubleshooting-mon/#known-limitations).

### Prerequisites

[toolbox](./ceph-toolbox.md) is deployed.

### Run the recovering script

Run [this script](https://github.com/rook/rook/blob/master/tests/scripts/restore-mon-quorum-using-osds.sh) as follows.

```console
./tests/scripts/restore-mon-quorum-using-osds.sh -f <cluster's fsid> -n rook-ceph -b ./backup -t ./tmp
```

If this script fails, run the following script to restore the original state.

```console
./tests/scripts/restore-mon-quorum-using-osds.sh -r -f <cluster's fsid> -n rook-ceph -b ./backup
```

### Check if the configuration is correct

> Some configurations might be lost after restoring cluster. Reconfigure them if necessary.

Author: It doesn't describe in Ceph's official document. I'll add a description which let users to backup Ceph's configuration before running this script and restore them if lost.
